File: test-suite-jackson/src/test/java/example/micronaut/GraphQLJacksonSerializationTest.java
package example.micronaut;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.micronaut.configuration.graphql.GraphQLResponseBody;
import io.micronaut.context.ApplicationContext;
import io.micronaut.context.annotation.Property;
import io.micronaut.json.JsonMapper;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Collections;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

@Property(name = "jackson.serializationInclusion", value = "ALWAYS")
@MicronautTest(startApplication = false)
class GraphQLJacksonSerializationTest {

    @Test
    void serializeGraphQLResponseBody(JsonMapper mapper) throws IOException {
        Map<String, Object> specification = Map.of("foo", "bar");
        var response = new GraphQLResponseBody(specification);
        var expected = """
            {"foo":"bar"}""";
        assertEquals(expected, mapper.writeValueAsString(response));
    }

    @Test
    void serializeEmptyGraphQLResponseBody(JsonMapper mapper) throws IOException {
        Map<String, Object> specification = Map.of("foo", Map.of("bar", Collections.emptyList()));
        var response = new GraphQLResponseBody(specification);
        var expected = """
            {"foo":{"bar":[]}}""";
        assertEquals(expected, mapper.writeValueAsString(response));
    }

    @Test
    void testRawJackson() throws JsonProcessingException {
        ObjectMapper objectMapper = new ObjectMapper();
        String result = objectMapper.writeValueAsString(
            new GraphQLResponseBody(Map.of("data", "test")));
        assertEquals("""
            {"data":"test"}""", result);
    }

    @Test
    void testInjectedJackson() throws JsonProcessingException {
        try (ApplicationContext ctx = ApplicationContext.run()) {
            ObjectMapper objectMapper = ctx.getBean(ObjectMapper.class);
            String result = objectMapper.writeValueAsString(
                new GraphQLResponseBody(Map.of("data", "test")));
            assertEquals("""
            {"data":"test"}""", result);
        }
    }
}

File: docs-examples/hello-world-kotlin/src/test/kotlin/io/micronaut/graphql/docs/GraphQLJacksonSerializationTest.kt
package io.micronaut.graphql.docs

import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import com.fasterxml.jackson.databind.ObjectMapper
import io.micronaut.configuration.graphql.GraphQLResponseBody
import org.junit.jupiter.api.Disabled

@MicronautTest
class GraphQLJacksonSerializationTest {

    @Inject
    lateinit var mapper: ObjectMapper

    @Disabled("https://github.com/micronaut-projects/micronaut-core/pull/9872")
    @Test
    fun serializeGraphQLResponseBody() {
        val specification = mapOf("foo" to "bar")
        val response = GraphQLResponseBody(specification)
        val expected = """{"foo":"bar"}"""
        val m = ObjectMapper()
        Assertions.assertEquals(expected, m.writeValueAsString(response))
        Assertions.assertEquals(expected, mapper.writeValueAsString(response))
    }
}
